core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -231,33 +226,26 @@ private object RTS {
       }
     }
 
-    private final def collectDefect[E, A](e: ExitResult[E, A]): Option[List[Throwable]] =
-      e match {
-        case ExitResult.Terminated(ts @ _ :: _) => Some(ts)
-        case ExitResult.Failed(_, ts @ _ :: _)  => Some(ts)
-        case _                                  => None
-      }
-
     /**
-     * Catches an exception, returning a (possibly null) finalizer action that
-     * must be executed. It is painstakingly *guaranteed* that the stack will be
-     * empty in the sole case the exception was not caught by any exception
-     * handler—i.e. the exceptional case.
+     * Empties the stack, collecting all finalizers and coalescing them into an
+     * action that produces a list (possibly empty) of errors during finalization.
+     * If needed, catch exceptions and apply redeem error handling.
      */
-    final def catchError: IO[Nothing, Option[List[Throwable]]] = {
-      var errorHandler: Any => IO[Any, Any]               = null
-      var finalizer: IO[Nothing, Option[List[Throwable]]] = null
+    final def unwindStack(catchError: Boolean): IO[Nothing, Option[Cause[Nothing]]] = {
+      var errorHandler: Any => IO[Any, Any]              = null
+      var finalizer: IO[Nothing, Option[Cause[Nothing]]] = null
 
       // Unwind the stack, looking for exception handlers and coalescing
       // finalizers.
       while ((errorHandler eq null) && !stack.isEmpty) {
         stack.pop() match {
-          case a: IO.Redeem[_, _, _, _] =>
+          case a: IO.Redeem[_, _, _, _] if catchError =>
             errorHandler = a.err.asInstanceOf[Any => IO[Any, Any]]
           case f0: Finalizer =>
-            val f: IO[Nothing, Option[List[Throwable]]] = fork(f0.finalizer, _ => IO.unit).observe.map(collectDefect)
+            val f: IO[Nothing, Option[Cause[Nothing]]] =
+              fork(f0.finalizer, _ => IO.unit).observe.map((a: ExitResult[Nothing, Unit]) => a.fold(_ => None, Some(_)))
```

Comment:
Worth adding a `cause: Option[Cause[E]]` onto `ExitResult`, so you can write this line as:

```scala
fork(f0.finalizer, _ => IO.unit).observe.map(_.cause)
```

Further, I'm not sure why the finalizer is being forked. It's probably a holdover from when we needed to do that in order to capture all failures. There's a better way to write that now:

```scala
f0.finalizer.sandbox.redeemPure[Option[Cause[Nothing]]](Some(_), None)
```

This will have the wonderful benefit of making all finalizers synchronous. 🎉 🕺 

@regiskuckaertz See any problems with this?

Reply:
Nope, haha that's perfect 💯

Reply:
Note, this PR does not change the run/sandbox encoding yet.

Reply:
Can you try implementing #360? It will just be a couple lines and may fix whatever issue is causing the failures.

Reply:
/cc @ghostdogpr

Reply:
I implemented `run` and it fixed the issue indeed!
Looking at `raceWith` now before I commit all.

Reply:
Don’t worry about that one, it is more involved and we can fix it in that ticket.

Reply:
Okay!